release/src/router/freeradius-server-3.0.0/man/man5/rlm_attr_filter.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_attr_filter 5 "27 June 2013" "" "FreeRADIUS Module"
.SH NAME
rlm_attr_filter \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_attr_filter\fP module exists for filtering certain
attributes and values in received ( or transmitted ) radius packets.
It gives the server a flexible framework to filter the attributes we
send to or receive from home servers or NASes.  This makes sense, for
example, in an out-sourced dialup situation to various policy
decisions, such as restricting a client to certain ranges of
Idle-Timeout or Session-Timeout.
.PP
Filter rules are normally defined and applied on a per-realm basis,
where the realm is anything that is defined and matched based on the
configuration of the \fIrlm_realm\fP module.  Filter rules can
optionally be applied using another attribute, by editing the
\fIkey\fP configuration for this module.
.PP
In 2.0.1 and earlier versions, the "accounting" section filtered the
Accounting-Request, even though it was documented as filtering the
response.  This issue has been fixed in version 2.0.2 and later
versions.  The "preacct" section may now be used to filter
Accounting-Request packets.  The "accounting" section now filters
Accounting-Response packets.  Administrators using "attr_filter" in
the "accounting" section SHOULD move the reference to "attr_filter"
from "accounting" to "preacct".
.PP
The file that defines the attribute filtering rules follows a similar
syntax to the \fIusers\fP file.  There are a few differences however:
.PP
.DS
    There are no check-items allowed other than the name of the key.
.PP
    There can only be a single DEFAULT entry.
.PP
The rules for each entry are parsed to top to bottom, and an
attribute must pass *all* the rules which affect it in order to
make it past the filter.  Order of the rules is important.
The operators and their purpose in defining the rules are as
follows:
.TP
.B =   
THIS OPERATOR IS NOT ALLOWED.  If used, and warning message is
printed and it is treated as ==
.TP
.B :=  
Set, this attribute and value will always be placed in the
output A/V Pairs.  If the attribute exists, it is overwritten.
.TP
.B  ==  
Equal, value must match exactly.
.TP
.B  =*  
Always Equal, allow all values for the specified attribute.
.TP
.B    !*  
Never Equal, disallow all values for the specified attribute.
( This is redundant, as any A/V Pair not explicitly permitted
will be dropped ).
.TP
.B    !=  
Not Equal, value must not match.
.TP
.B    >=  
Greater Than or Equal
.TP
.B    <=  
Less Than or Equal
.TP
.B    >   
Greater Than
.TP
.B    <   
Less Than
.PP
If regular expressions are enabled the following operators are
also possible.  ( Regular Expressions are included by default
unless your system doesn't support them, which should be rare ).
The value field uses standard regular expression syntax.
.PP
.TP
.B    =~  
Regular Expression Equal
.TP
.B    !~  
Regular Expression Not Equal
.PP
See the default \fI/etc/raddb/attrs\fP for working examples of
sample rule ordering and how to use the different operators.
.DE
.PP
The configuration items are:
.IP file
This specifies the location of the file used to load the filter rules.
This file is used to filter the accounting response, packet before it
is proxied, proxy response from the home server, or our response to
the NAS.
.IP key
Usually %{Realm} (the default).  Can also be %{User-Name}, or other
attribute that exists in the request.  Note that the module always
keys off of attributes in the request, and NOT in any other packet.
.IP relaxed
If set to 'yes', then attributes which do not match any filter rules
explicitly, will also be allowed. This behaviour may be overridden 
for an individual filter block using the Relax-Filter check item.
The default for this configuration item is 'no'.
.PP
.SH SECTIONS
.IP preacct
Filters Accounting-Request packets.
.IP accounting
Filters Accounting-Response packets.
.IP pre-proxy
Filters Accounting-Request or Access-Request packets prior to proxying
them.
.IP post-proxy
Filters Accounting-Response, Access-Accept, Access-Reject, or
Access-Challenge responses from a home server.
.IP authorize
Filters Access-Request packets.
.IP post-auth
Filters Access-Accept or Access-Reject packets.
.PP
.SH FILES
.I /etc/raddb/radiusd.conf
.I /etc/raddb/filter/*
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5)
.SH AUTHOR
Chris Parker, cparker@segv.org



release/src/router/freeradius-server-3.0.0/src/modules/rlm_attr_filter/rlm_attr_filter.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License, version 2 if the
 *   License as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_attr_filter.c
 * @brief Filter the contents of a list, allowing only certain attributes.
 *
 * @copyright (C) 2001,2006 The FreeRADIUS server project
 * @copyright (C) 2001 Chris Parker <cparker@starnetusa.net>
 */
RCSID("$Id$")

#include	<freeradius-devel/radiusd.h>
#include	<freeradius-devel/modules.h>
#include	<freeradius-devel/rad_assert.h>

#include	<sys/stat.h>

#include	<ctype.h>
#include	<fcntl.h>
#include	<limits.h>


/*
 *	Define a structure with the module configuration, so it can
 *	be used as the instance handle.
 */
typedef struct rlm_attr_filter {
	char		*filename;
	char		*key;
	int		relaxed;
	PAIR_LIST	*attrs;
} rlm_attr_filter_t;

static const CONF_PARSER module_config[] = {
	{ "attrsfile",     PW_TYPE_FILE_INPUT | PW_TYPE_DEPRECATED,
	  offsetof(rlm_attr_filter_t, filename), NULL, NULL},
	{ "filename",     PW_TYPE_FILE_INPUT | PW_TYPE_REQUIRED,
	  offsetof(rlm_attr_filter_t, filename), NULL, NULL},
	{ "key",     PW_TYPE_STRING_PTR,
	  offsetof(rlm_attr_filter_t, key), NULL, "%{Realm}" },
	{ "relaxed",    PW_TYPE_BOOLEAN,
	  offsetof(rlm_attr_filter_t, relaxed), NULL, "no" },
	{ NULL, -1, 0, NULL, NULL }
};

static void check_pair(VALUE_PAIR *check_item, VALUE_PAIR *reply_item,
		      int *pass, int *fail)
{
	int compare;

	if (check_item->op == T_OP_SET) return;

	compare = paircmp(check_item, reply_item);
	if (compare == 1) {
		++*(pass);
	} else {
		++*(fail);
	}

	return;
}


static int attr_filter_getfile(TALLOC_CTX *ctx, char const *filename, PAIR_LIST **pair_list)
{
	vp_cursor_t cursor;
	int rcode;
	PAIR_LIST *attrs = NULL;
	PAIR_LIST *entry;
	VALUE_PAIR *vp;

	rcode = pairlist_read(ctx, filename, &attrs, 1);
	if (rcode < 0) {
		return -1;
	}

	/*
	 * Walk through the 'attrs' file list.
	 */

	entry = attrs;
	while (entry) {
		entry->check = entry->reply;
		entry->reply = NULL;

		for (vp = paircursor(&cursor, &entry->check);
		     vp;
		     vp = pairnext(&cursor)) {
		    /*
		     * If it's NOT a vendor attribute,
		     * and it's NOT a wire protocol
		     * and we ignore Fall-Through,
		     * then bitch about it, giving a good warning message.
		     */
		     if ((vp->da->vendor == 0) &&
			 (vp->da->attr > 0xff) &&
			 (vp->da->attr > 1000)) {
			WDEBUG("[%s]:%d Check item \"%s\"\n\tfound in filter list for realm \"%s\".\n",
			       filename, entry->lineno, vp->da->name, entry->name);
		    }
		}

		entry = entry->next;
	}

	*pair_list = attrs;
	return 0;
}


/*
 *	(Re-)read the "attrs" file into memory.
 */
static int mod_instantiate(UNUSED CONF_SECTION *conf, void *instance)
{
	rlm_attr_filter_t *inst = instance;
	int rcode;

	rcode = attr_filter_getfile(inst, inst->filename, &inst->attrs);
	if (rcode != 0) {
		ERROR("Errors reading %s", inst->filename);

		return -1;
	}

	return 0;
}


/*
 *	Common attr_filter checks
 */
static rlm_rcode_t attr_filter_common(void *instance, REQUEST *request, RADIUS_PACKET *packet)
{
	rlm_attr_filter_t *inst = instance;
	VALUE_PAIR	*vp;
	vp_cursor_t	input, check, out;
	VALUE_PAIR	*input_item, *check_item, *output;
	PAIR_LIST	*pl;
	int		found = 0;
	int		pass, fail = 0;
	char const	*keyname = NULL;
	char		buffer[256];

	if (!packet) return RLM_MODULE_NOOP;

	if (!inst->key) {
		VALUE_PAIR	*namepair;

		namepair = pairfind(request->packet->vps, PW_REALM, 0, TAG_ANY);
		if (!namepair) {
			return (RLM_MODULE_NOOP);
		}
		keyname = namepair->vp_strvalue;
	} else {
		int len;

		len = radius_xlat(buffer, sizeof(buffer), request, inst->key, NULL, NULL);
		if (len < 0) {
			return RLM_MODULE_FAIL;
		}
		if (len == 0) {
			return RLM_MODULE_NOOP;
		}
		keyname = buffer;
	}

	/*
	 *	Head of the output list
	 */
	output = NULL;
	paircursor(&out, &output);

	/*
	 *      Find the attr_filter profile entry for the entry.
	 */
	for (pl = inst->attrs; pl; pl = pl->next) {
		int fall_through = 0;
		int relax_filter = inst->relaxed;

		/*
		 *  If the current entry is NOT a default,
		 *  AND the realm does NOT match the current entry,
		 *  then skip to the next entry.
		 */
		if ((strcmp(pl->name, "DEFAULT") != 0) &&
		    (strcmp(keyname, pl->name) != 0))  {
		    continue;
		}

		RDEBUG2("Matched entry %s at line %d", pl->name, pl->lineno);
		found = 1;

		for (check_item = paircursor(&check, &pl->check);
		     check_item;
		     check_item = pairnext(&check)) {
			if (!check_item->da->vendor &&
			    (check_item->da->attr == PW_FALL_THROUGH) &&
				(check_item->vp_integer == 1)) {
				fall_through = 1;
				continue;
			}
			else if (!check_item->da->vendor &&
				 check_item->da->attr == PW_RELAX_FILTER) {
				relax_filter = check_item->vp_integer;
				continue;
			}

			/*
			 *    If it is a SET operator, add the attribute to
			 *    the output list without checking it.
			 */
			if (check_item->op == T_OP_SET ) {
				vp = paircopyvp(packet, check_item);
				if (!vp) {
					goto error;
				}
				radius_xlat_do(request, vp);
				pairinsert(&out, vp);
			}
		}

		/*
		 *	Iterate through the input items, comparing
		 *	each item to every rule, then moving it to the
		 *	output list only if it matches all rules
		 *	for that attribute.  IE, Idle-Timeout is moved
		 *	only if it matches all rules that describe an
		 *	Idle-Timeout.
		 */
		for (input_item = paircursor(&input, &packet->vps);
		     input_item;
		     input_item = pairnext(&input)) {
			/* reset the pass,fail vars for each reply item */
			pass = fail = 0;

			/*
			 *	reset the check_item pointer to
			 *	beginning of the list
			 */
			for (check_item = pairfirst(&check);
			     check_item;
			     check_item = pairnext(&check)) {
				/*
				 *	Vendor-Specific is special, and
				 *	matches any VSA if the comparison
				 *	is always true.
				 */
				if ((check_item->da->attr == PW_VENDOR_SPECIFIC) && (input_item->da->vendor != 0) &&
				    (check_item->op == T_OP_CMP_TRUE)) {
					pass++;
					continue;
				}

				if (input_item->da->attr == check_item->da->attr) {
					check_pair(check_item, input_item, &pass, &fail);
				}
			}

			/*
			 *  Only move attribute if it passed all rules,
			 *  or if the config says we should copy unmatched
			 *  attributes ('relaxed' mode).
			 */
			if (fail == 0 && (pass > 0 || relax_filter)) {
				if (!pass) {
					RDEBUG3("Attribute (%s) allowed by relaxed mode", input_item->da->name);
				}
				vp = paircopyvp(packet, input_item);
				if (!vp) {
					goto error;
				}
				pairinsert(&out, vp);
			}
		}

		/* If we shouldn't fall through, break */
		if (!fall_through) {
			break;
		}
	}

	/*
	 *	No entry matched.  We didn't do anything.
	 */
	if (!found) {
		rad_assert(!output);
		return RLM_MODULE_NOOP;
	}

	/*
	 *	Replace the existing request list with our filtered one
	 */
	pairfree(&packet->vps);
	packet->vps = output;

	if (request->packet->code == PW_AUTHENTICATION_REQUEST) {
		request->username = pairfind(request->packet->vps, PW_STRIPPED_USER_NAME, 0, TAG_ANY);
		if (!request->username)
			request->username = pairfind(request->packet->vps, PW_USER_NAME, 0, TAG_ANY);
		request->password = pairfind(request->packet->vps, PW_USER_PASSWORD, 0, TAG_ANY);
	}

	return RLM_MODULE_UPDATED;

	error:
	pairfree(&output);
	return RLM_MODULE_FAIL;
}

#define RLM_AF_FUNC(_x, _y) static rlm_rcode_t mod_##_x(void *instance, REQUEST *request) \
			{ \
				return attr_filter_common(instance, request, request->_y); \
			}

RLM_AF_FUNC(authorize, packet)
RLM_AF_FUNC(post_auth, reply)

RLM_AF_FUNC(preacct, packet)
RLM_AF_FUNC(accounting, reply)

#ifdef WITH_PROXY
RLM_AF_FUNC(pre_proxy, proxy)
RLM_AF_FUNC(post_proxy, proxy_reply)
#endif

#ifdef WITH_COA
RLM_AF_FUNC(recv_coa, packet)
RLM_AF_FUNC(send_coa, reply)
#endif

/* globally exported name */
module_t rlm_attr_filter = {
	RLM_MODULE_INIT,
	"attr_filter",
	RLM_TYPE_CHECK_CONFIG_SAFE | RLM_TYPE_HUP_SAFE,   	/* type */
	sizeof(rlm_attr_filter_t),
	module_config,
	mod_instantiate,	/* instantiation */
	NULL,			/* detach */
	{
		NULL,			/* authentication */
		mod_authorize,	/* authorization */
		mod_preacct,	/* pre-acct */
		mod_accounting,	/* accounting */
		NULL,			/* checksimul */
#ifdef WITH_PROXY
		mod_pre_proxy,	/* pre-proxy */
		mod_post_proxy,	/* post-proxy */
#else
		NULL, NULL,
#endif
		mod_post_auth	/* post-auth */
#ifdef WITH_COA
		,
		mod_recv_coa,
		mod_send_coa
#endif
	},
};

